staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filter.go

```diff
@@ -0,0 +1,273 @@
+		if fs == nil { // reject
+			metrics.AddReject("<none>", "non-match")
+			klog.V(7).Infof("Rejecting requestInfo=%#+v, userInfo=%#+v because no FlowSchema matched", requestDigest.RequestInfo, requestDigest.User)
+			return false, func() {}
+		}
+		plName := fs.Spec.PriorityLevelConfiguration.Name
+
+		// 2. early out for exempt
+		ps := rmState.priorityLevelStates[plName]
+		if ps.config.Type == rmtypesv1alpha1.PriorityLevelEnablementExempt {
+			klog.V(7).Infof("Serving requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s without delay", requestDigest.RequestInfo, requestDigest.User, fs.Name, plName)
+			startExecutionTime := time.Now()
+			return true, func() {
+				metrics.ObserveExecutionDuration(plName, fs.Name, time.Now().Sub(startExecutionTime))
+			}
+		}
+
+		// 3. computing hash
+		flowDistinguisher := requestDigest.ComputeFlowDistinguisher(fs.Spec.DistinguisherMethod)
+		hashValue := hashFlowID(fs.Name, flowDistinguisher)
+
+		// 4. queuing
+		tryAnother, execute, afterExecute := ps.queues.Wait(ctx, hashValue, requestDigest.RequestInfo, requestDigest.User)
+		if tryAnother {
+			klog.V(5).Infof("Request requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s landed in timing splinter, re-classifying", requestDigest.RequestInfo, requestDigest.User, fs.Name, plName)
+			continue
+		}
+
+		// 5. execute or reject
+		metrics.ObserveWaitingDuration(plName, fs.Name, strconv.FormatBool(execute), time.Now().Sub(startWaitingTime))
+		if !execute {
+			klog.V(7).Infof("Rejecting requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s after fair queuing", requestDigest.RequestInfo, requestDigest.User, fs.Name, plName)
+			return false, func() {}
+		}
+		klog.V(7).Infof("Serving requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s after fair queuing", requestDigest.RequestInfo, requestDigest.User, fs.Name, plName)
+		startExecutionTime := time.Now()
+		return execute, func() {
+			metrics.ObserveExecutionDuration(plName, fs.Name, time.Now().Sub(startExecutionTime))
+			afterExecute()
+		}
+	}
+}
+
+func (rmState *requestManagerState) pickFlowSchema(rd RequestDigest) *rmtypesv1alpha1.FlowSchema {
+	for _, flowSchema := range rmState.flowSchemas {
+		if matchesFlowSchema(rd, flowSchema) {
+			return flowSchema
+		}
+	}
+	return nil
+}
+
+// ComputeFlowDistinguisher extracts the flow distinguisher according to the given method
+func (rd RequestDigest) ComputeFlowDistinguisher(method *rmtypesv1alpha1.FlowDistinguisherMethod) string {
+	if method == nil {
+		return ""
+	}
+	switch method.Type {
+	case rmtypesv1alpha1.FlowDistinguisherMethodByUserType:
+		return rd.User.GetName()
+	case rmtypesv1alpha1.FlowDistinguisherMethodByNamespaceType:
+		return rd.RequestInfo.Namespace
+	default:
+		// this line shall never reach
+		panic("invalid flow-distinguisher method")
+	}
+}
+
+// HashFlowID hashes the inputs into 64-bits
+func hashFlowID(fsName, fDistinguisher string) uint64 {
+	return crc64.Checksum([]byte(fsName+fDistinguisher), crc64.MakeTable(crc64.ECMA))
```

Comment:
My intuition is that it's preferable to use a cryptographic quality hash function. It costs us little to nothing, and it would make this system very difficult to manipulate. For example, given this function as written, I think I can chose a namespace or service account name such that it collides with another name I know about, and use this to starve that other service account / namespace. I am not sure how feasible other aspects of such an attack would be. Most attacks assemble a few different problems in a surprising way; therefore, it's bad karma to leave puzzle pieces like this around.

SHA256 is my default hash function, there's $billions to be made if you can break it (bitcoin), so I'm very confident it's not broken.

My next problem here is the string concatenation `fsName+fDistinguisher`; it does an extra allocation. But it also means that "ab" +"abac" is the same as "abab" + "ac". It's better to either length-prefix the value or hash in a separator.

Reply:
Oops, @lavalamp beat me to it :)

> My next problem here is the string concatenation fsName+fDistinguisher; it does an extra allocation.

FWIW, this is pretty easy to work around if it turns into a performance problem since just about every hashing API allows for incremental hashing.

Reply:
Yeah that was my (poorly stated) point--we can probably pay for most of a "real" hashing function just by removing that concatenation.

Reply:
We discussed whether to use a cryptographic hash months ago when we started this, and decided not to.  Perhaps I will do some benchmarking.
I suggested bumming out the concatenation long ago.

Reply:
next push eschews the concatenation and includes a delimiter.